tel2puml/find_unique_graphs/otel_ingestion/otel_data_holder.py

```
@@ -2,19 +2,171 @@
+
+
+class SQLDataHolder(DataHolder):
+    """A class to handle saving data in SQL databases using SQLAlchemy."""
+
+    def __init__(self, config: SQLDataHolderConfig) -> None:
+        """Constructor method.
+
+        :param config: Configuration parameters.
+        :type config: :class: `SQLDataHolderConfig`
+        """
+        super().__init__()
+        self.node_models_to_save: list[NodeModel] = []
+        self.node_relationships_to_save: list[dict[str, str]] = []
+        self.batch_size: int = config["batch_size"]
+        self.engine: Engine = create_engine(config["db_uri"], echo=False)
+        self.session: Session = Session(bind=self.engine)
+        self.base: Base = Base()
+        self.create_db_tables()
+
+    def create_db_tables(self) -> None:
+        """Method to create the database tables based on the defined models."""
+
+        self.base.metadata.create_all(self.engine)
+
+    def _save_data(self, otel_event: OTelEvent) -> None:
+        """Method for batching and saving OTel data to SQL database.
+
+        :param otel_event: An OTelEvent object.
+        :type otel_event: :class: `OTelEvent`
+        """
+        node_model = self.convert_otel_event_to_node_model(otel_event)
+
+        self.node_models_to_save.append(node_model)
+        self.add_node_relations(otel_event)
+
+        if len(self.node_models_to_save) >= self.batch_size:
+            self.commit_batched_data_to_database()
+
+    def commit_batched_data_to_database(self) -> None:
+        """Method to commit batched node models, and their relationships to
+        a SQL database.
+        """
+
+        try:
+            self.batch_insert_node_models()
+            self.batch_insert_node_associations()
+            # Reset batch
+            self.node_models_to_save = []
+            self.node_relationships_to_save = []
+        except Exception as e:
+            self.session.rollback()
+            raise Exception(f"Unexpected error during data saving: {e}")
```

If we can't be more specific can we just re-raise the exception instead i.e. raise e